content/en/docs/architecture/timeouts.md

```diff
@@ -73,8 +106,54 @@ ref:
     Runs the end-to-end suite published by org/repo.
 ```
 
-## FAQ
+{{< alert title="Warning" color="warning" >}}
+The `pod.spec.activeDeadlineSeconds` setting on a `Pod` only implicitly bounds the amount of time that a `Pod` executes for on a Kubernetes cluster. The active deadline begins at the first moment that a `kubelet` acknowledges the `Pod`, which is after it is scheduled to a specific node but before it pulls images, sets up a container sandbox, _etc_. It is therefore possible to exceed the active deadline without ever having a container in the `Pod` execute. Please see the [API documentation](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#podspec-v1-core) for more details.
+{{< /alert >}}
+
+## How Interruptions May Be Handled
+
+Two main approaches exist to handling interruptions for a test process: first, the test process itself may listen for and handle `SIGTERM`; second, `post` steps may be declared in a test `workflow` to be run after an interruption occurs. The first approach is most useful when relevant state for responding to the interrupt exists only in the test process itself, and the response is fairly short. This approach has the downside of requiring complex test process code and signal handling implementation. The second approach is suggested as it is more robust and tunable. In this approach, state needed to respond to the interrupt should be stored in the [`${SHARED_DIR}`](/docs/architecture/step-registry/#sharing-data-between-steps) for use by the `post` step. The `post` step may be marked as [best-effort](/docs/architecture/step-registry/#marking-post-steps-best-effort) if it only gathers artifacts or cleans up resources. Examples of both approaches follow.
+
+### Handling `SIGTERM` In A Test Process
+
+An individual test process configured as a step in a job workflow may handle `SIGTERM` itself to gracefully exit and produce all logs and artifacts necessary. Handling interruption signals within a test process is useful when test process state is necessary to correctly handle interruption. Note that the grace period for a step must be long enough to both generate any artifacts or logs that need to be uploaded _and_ for the upload to occur. Here's an example of a step configuration that handles `SIGTERM` itself:
+
+```yaml
+ref:
+  as: my-test
+  from: src
+  commands: my-test-commands.sh
+  resources:
+    requests:
+      cpu: 1000m
+      memory: 100Mi
+  termination_grace_period_seconds: 120
+  documentation: |-
+    Runs the test suite from source code.
+```
+
+Note that the step declares a 120s termination grace period, which is much larger than the default of 30s. The step commands might look like:
+
+```shell
+#!/usr/bin/env bash
+
+set -o errexit
+set -o nounset
+set -o pipfail
+
+function cleanup() {
+  echo "Validated up to ${current}!"
+}
+
+trap cleanup SIGTERM
+
+current=0
```

Comment:
micro-nit: probably want `current` defined before `cleanup`, or at least before the `trap` to avoid `Validated up to ` with no number if the `TERM` lands right after the `trap` gets the handler in place.

Reply:
Also, while I approve of the simplicity of this contrived example, we do have a lot of steps [like the installer step][1] that use a pattern like:

```shell
#!/bin/bash

set -o nounset
set -o errexit
set -o pipefail

trap 'CHILDREN=$(jobs -p); if test -n "${CHILDREN}"; then kill ${CHILDREN} && wait; fi' TERM

some-slow-thing &
wait "$!"

and-then-more-stuff  # without this, you could just: exec some-slow-thing
```

to get TERM propagation into the slow children.  Several examples demonstrating the approach and some pitfalls it avoids are in the openshift/release@4472ace1205dc9c15f59c31ee63db99723e9ba39 commit message.

[1]: https://github.com/openshift/release/blob/160d88776eedf0a9e4fbbc8ab9d5805ffc42d9d5/ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh

Reply:
Yeah, when job control is not enabled as its the default on noninteractive shells, you need to start the children in a way that they do not block the shell, otherwise it will not react to signals for as long as its blocked by a child.

The simpler approach to cope with this is to enable it via `set -m` though, rather than the whole forking dance